pybamm/experiment/experiment.py

```diff
@@ -73,7 +73,7 @@ def __init__(
         for cycle in operating_conditions:
             # Check types and convert to list
             if not isinstance(cycle, tuple):
-                cycle = (cycle,)
+                cycle = (cycle,)  # type: ignore[assignment]
```

Comment:
Is it possible to avoid these type of ignore comments for the typing?

Reply:
These comments are used by mypy when static type checking. I've done my best to use them as little as possible, but in some cases the amount of rewriting required to make mypy happy would make the code less readable than having the comment (in my opinion). If we decide not to use a type checker, they can be removed.

Reply:
This instance isn't actually a brilliant example as I've managed to get rid of the type error; however there are ~ 10 ignore statements I've had to leave on a line-by-line basis. Mypy raises an error for each `# type: ignore[]` comment even with strict mode turned off, but if we don't enforce mypy as having to pass in the CI then we could chose to use it as a 'todo' list of sorts and remove the `# type: ignore` comments.

The remaining errors are split about 50/50 between difficulties integrating mypy into heavily dynamically-typed areas (e.g. it won't let you re-define a variable with a different type before you've used it, even with the --allow-redefinition flag https://mypy.readthedocs.io/en/latest/command_line.html#miscellaneous-strictness-flags), and areas where pybamm would require more significant re-writing. For example in `serialise.py`, where mypy complains that `BaseModel` doesn't have a `bounds` attribute. This is true at the point where BaseModel is defined, but the attribute is added once the model is discretised; as that particular function only takes a discretised model, the code works properly but I haven't found a way to flag to mypy that this is the case. There are multiple ways to fix this, including creating a `DiscretisedModel` class or editing the `__init__` of `BaseModel` to include attributes that are added later, but that's a lot of work & discussion that shouldn't be done here.